### ISSUE TITLE
F# Option (de)serialization is broken

### ISSUE THREAD

**Issue by ca-ta** [dotnet/orleans#1056]
Discovered while attempting to upgrade to the latest version (82c1eb833224f002ee0611ae7b00c925d6ee139d). I could definitely use some help in solving this - I am not at all familiar with the serialization mechanism.

Serialization of the F# Option type seems broken - I'll submit  a repro PR shortly.

**Comment by ReubenBond** [comment]
If a grain interface returns an `Option<T>`, does everything work? From the repro PR, I you never statically specify `Option` anywhere in the interface definition, only at runtime.

I'm not saying that returning `Option<T>` should be required, I just want to understand the scope of the problem: i.e, is serializer generation broken for F# or just not being invoked because it isn't specified.

Once we establish where the issue lies, we can work on fixing it.

**Comment by ca-ta** [comment]
@ReubenBond - thank you for looking into this!

Based on the failure call stacks, Orleans serializers do get generated, registered and invoked. 
Note also that only a subset of the tests in #1062 (which I think will be merged once @jthelin vets it) fail. 

There seem to be two distinct failure modes:
 1. Calling a grain interface method that takes an ```Option<T>``` when the parameter is ```None``` fails at serialization time - the reason seems to be that ```None``` is represented as ```null```. 
 2. Calling a grain interface method that returns an F# record that has an ```Option<T>``` [value](https://msdn.microsoft.com/EN-US/library/dd233184(v=VS.140,d=hv.2).aspx) has the following behavior:
  1. if the value is ```None```, the call succeeds.
  2. If the Value is ```Some x```, the ```Option<T>``` record value deserialization fails with ```System.Security.VerificationException: Operation could destabilize the runtime.```, thrown  by either  ```concrete.Deserialize``` or ```concrete.DeepCopy``` (SerializationManager.cs, lines 491/496), depending of whether the record is decorated with the ```Immutable``` attribute or not.
    * the corresponding tests are ```FSharpGrains_Ping_GenericRecord_ofIntOption_Some``` and ```FSharpGrains_Ping_Record_ofIntOption_Some_WithNoAttributes```

Interestingly enough, if one runs the corresponding roundtrip serialization unit test for records with ```Option<T>``` record value in isolation, *that* unit test (```SerializationTests_FSharp_Record_ofIntOption_Some```) succeeds. It appears that that ```SerializationTests_FSharp_Record_ofIntOption_Some```succeeds in isolation because an Orleans serializer is not generated in that scenario (due, I think, to the very fact that you have pointed out: there are no concrete interfaces declared that use Option<T>).

So, to recap, I think we are looking at 2 distinct issues:
1. ```None``` breaks the serialization because it is represented as null
2. F# records that contain  ```Option<T>``` values break the Orleans generated serializers if and only if the value is ```Some``` (None seems to roundrip OK). 

I hope this helps - let me know if you need me to create more tests exercising these specific issues - I will happily do so once #1062 is merged.

**Comment by ReubenBond** [comment]
Thank you for this, I am fixing the code. I've added `[KnownType]` and `[KnownAssembly]`, which for type safety reasons can take a `typeof(x)` or an assembly name.

I believe `PingTest<T>` is broken. Unless I'm reading the (decompiled) code for `FSharpOption<T>` wrong, `None` is supposed to be represented as `null`:
```C#
/// <summary>
/// Create an option value that is a 'None' value.
/// </summary>
[CompilerGenerated]
[DebuggerNonUserCode]
[DebuggerBrowsable(DebuggerBrowsableState.Never)]
public static FSharpOption<T> None
{
    [CompilationMapping(SourceConstructFlags.UnionCase, 0)] get
    {
    return (FSharpOption<T>) null;
    }
}
```

Your `PingTest<T>` was performing equality checks using `Assert.IsTrue(output.Equals(input))`, which will fail with a `NullReferenceException` if `output` is null. I changed it to `Assert.AreEqual(input, output)` and all tests pass.

Currently I have to add `[<assembly: KnownAssembly(typedefof<unit option>)>]` to the F# project so that code generation is performed for the core F# library.

We can consider adding an `Orleans.FSharp.dll` which includes pre-generated code for the core F# libraries.

**Comment by ca-ta** [comment]
@ReubenBond - awesome, thank you!

The ``` PingTest<T> ``` test was obviously broken - thank you for spotting that and for the fix.  

Including a library with pre-generated code for the core F# libraries would be great - I would happily help with that. 
Would it suffice for the ``` Orleans.FSharp.dll  ``` dll to reference the F# core library and include an ``` [<assembly: KnownAssembly("FSharp.Core.dll")>] ``` (or multiple of the type variant) attribute, or is there more to it? Sorry if I'm being thick - I have yet to find the time to understand the new code generation mechanism. Any pointers you can share would be greatly appreciated.

**Comment by ReubenBond** [comment]
> Would it suffice for the Orleans.FSharp.dll dll to reference the F# core library and include an [<assembly: KnownAssembly("FSharp.Core.dll")>] (or multiple of the type variant) attribute, or is there more to it?

*Almost*, except that it should be a C# project so we can enable compile-time code generation, since that's the only kind we support now.

**Comment by ca-ta** [comment]
Regarding the project being C#  - yes, that's understood.
I will submit a PR and additional treats after Thanksgiving.

Thanks!

**Comment by sergeybykov** [comment]
Resolved by https://github.com/dotnet/orleans/pull/1095.
